clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/SimpleClusteredTierClientEntity.java

```diff
@@ -0,0 +1,268 @@
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client.internal.store;
+
+import org.ehcache.clustered.client.config.TimeoutDuration;
+import org.ehcache.clustered.client.internal.Timeouts;
+import org.ehcache.clustered.client.internal.service.ClusteredTierException;
+import org.ehcache.clustered.client.internal.service.ClusteredTierValidationException;
+import org.ehcache.clustered.common.internal.ServerStoreConfiguration;
+import org.ehcache.clustered.common.internal.exceptions.ClusterException;
+import org.ehcache.clustered.common.internal.messages.EhcacheEntityMessage;
+import org.ehcache.clustered.common.internal.messages.EhcacheEntityResponse;
+import org.ehcache.clustered.common.internal.messages.EhcacheEntityResponse.Failure;
+import org.ehcache.clustered.common.internal.messages.EhcacheMessageType;
+import org.ehcache.clustered.common.internal.messages.EhcacheOperationMessage;
+import org.ehcache.clustered.common.internal.messages.EhcacheResponseType;
+import org.ehcache.clustered.common.internal.messages.LifeCycleMessageFactory;
+import org.ehcache.clustered.common.internal.messages.ReconnectMessage;
+import org.ehcache.clustered.common.internal.messages.ReconnectMessageCodec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terracotta.entity.EndpointDelegate;
+import org.terracotta.entity.EntityClientEndpoint;
+import org.terracotta.entity.EntityResponse;
+import org.terracotta.entity.InvokeFuture;
+import org.terracotta.entity.MessageCodecException;
+import org.terracotta.exception.EntityException;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * ClusteredTierClientEntity
+ */
+public class SimpleClusteredTierClientEntity implements InternalClusterTierClientEntity {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SimpleClusteredTierClientEntity.class);
+  private static final Set<EhcacheMessageType> GET_STORE_OPS = EnumSet.of(EhcacheMessageType.GET_STORE);
+
+  private final EntityClientEndpoint<EhcacheEntityMessage, EhcacheEntityResponse> endpoint;
+  private final LifeCycleMessageFactory messageFactory;
+  private final AtomicLong sequenceGenerator = new AtomicLong(0L);
+  private final Object lock = new Object();
+  private final ReconnectMessageCodec reconnectMessageCodec = new ReconnectMessageCodec();
+  private final Map<Class<? extends EhcacheEntityResponse>, List<ResponseListener<? extends EhcacheEntityResponse>>> responseListeners =
+    new ConcurrentHashMap<Class<? extends EhcacheEntityResponse>, List<ResponseListener<? extends EhcacheEntityResponse>>>();
+
+  private UUID clientId;
```

Comment:
Double checked, since the `SimpleClusteredTierClientEntity` is only ever used by other threads after being assigned to a final field in a `ServerStoreProxy` it is safe.